lib/pwnlib/elf/elf.rb

```diff
@@ -0,0 +1,210 @@
+      # @return [Boolean] Yes or not.
+      def pie?
+        @elf_file.elf_type == 'DYN'
+      end
+
+      # There's too many objects inside, let pry not so verbose.
+      # @return [nil]
+      def inspect
+        nil
+      end
+
+      private
+
+      def show_info
+        # TODO: Use logger?
+        puts checksec
+      end
+
+      # Get the dynamic tag with +type+.
+      # @return [ELFTools::Dynamic::Tag, nil]
+      def dynamic_tag(type)
+        dynamic = @elf_file.segment_by_type(:dynamic) || @elf.section_by_name('.dynamic')
+        return nil if dynamic.nil? # No dynamic present, might be static-linked.
+        dynamic.tag_by_type(type)
+      end
+
+      # Load got symbols
+      def load_got
+        @got = OpenStruct.new
+        sections_by_types(%i(rel rela)).each do |rel_sec|
+          symtab = @elf_file.section_at(rel_sec.header.sh_link)
+          next unless symtab.respond_to?(:symbol_at)
+          rel_sec.relocations.each do |rel|
+            symbol = symtab.symbol_at(rel.symbol_index)
+            next unless symbol
+            @got[symbol.name] = rel.header.r_offset
+          end
+        end
+      end
+
+      # Load all plt symbols.
+      def load_plt
+        # Unlike pwntools-python, which use unicorn emulating instructions to find plt(s).
+        # Here only use section information, which won't find any plt(s) when compile option '-Wl' is enabled.
+        #
+        # The implementation here same as python-pwntools 3.5, and supports i386 and amd64 only.
+        @plt = OpenStruct.new
+        plt_sec = @elf_file.section_by_name('.plt')
+        return if plt_sec.nil? # TODO: log.warn
+        rel_sec = @elf_file.section_by_name('.rel.plt') || @elf_file.section_by_name('.rela.plt')
+        return if rel_sec.nil? # -Wl enabled
+        symtab = @elf_file.section_at(rel_sec.header.sh_link)
+        address = plt_sec.header.sh_addr + 0x10 # magic offset, correct in i386/amd64
+        rel_sec.relocations.each do |rel|
+          symbol = symtab.symbol_at(rel.symbol_index)
+          @plt[symbol.name] = address
+          address += 0x10 # magic gap again
+        end
+      end
+
+      # Load all exist symbols.
+      def load_symbols
+        @symbols = OpenStruct.new
+        @elf_file.each_sections do |section|
+          next unless section.respond_to?(:symbols)
+          section.each_symbols do |symbol|
+            # Don't care symbols without name.
+            next if symbol.name.empty?
+            next if symbol.header.st_value.zero?
+            @symbols[symbol.name] = symbol.header.st_value
+          end
+        end
+      end
+
+      def sections_by_types(types)
+        types.map { |type| @elf_file.sections_by_type(type) }.flatten
+      end
+
+      def base_address
+        return 0 if pie?
+        # Find the min of PT_LOAD's p_vaddr
+        @elf_file.segments_by_type(:load)
+                 .map { |seg| seg.header.p_vaddr }
+                 .select { |addr| addr > 0 }
```

Comment:
There're some segments with `p_vaddr` equals zero
I'll survey why and add a comment for it